### ISSUE TITLE
Don't try to suspend tabs with the "window.onbeforeunload" event 

### ISSUE THREAD

**Issue by rodrigograca31** [deanoemcke/thegreatsuspender#272]
Some pages have code like this one:

```javascript
window.onbeforeunload = function() {
    return "Are you sure?";
};
```

When "The Great Suspender" tries to suspend them it fires the event.... And focus the page.... (because it's like an alert)

More information: http://www.w3schools.com/jsref/event_onbeforeunload.asp

Example: 
![](http://i.imgur.com/qEcsNK7.png)

**Comment by ajohnson23** [comment]
+1

**Comment by AndrewJDR** [comment]
Here's an example of a page that recently stole focus from me when it unloaded:
https://www.cambiaresearch.com/articles/15/javascript-char-codes-key-codes

I checked the event listeners registered to the page using dev tools. It has a number of unload events registered:
![unload-event-listeners](https://cloud.githubusercontent.com/assets/1588857/24016929/3b6c6654-0a4b-11e7-8beb-6ef0d2d6855e.png)

This super annoying when this happens, because sometimes I'm in the middle of typing something on a page that's buried somewhere in my dozens of open tabs, and I suddenly get switched away from when I was working on.

I'm actually happy to write a patch to fix this bug, but I'd like some suggestions on the best way to go about it. I'm guessing a lot of pages have unload event handlers, so simply not suspending such pages would probably result in *many* pages remaining unsuspended.

Does anyone know if there's a way from an extension to unbind some event handlers from a page? This way, we could just unbind all the unload event handlers before suspending the tab.

**Comment by AndrewJDR** [comment]
Going to call out @deanoemcke in case he has any suggestions... Thanks!

**Comment by rodrigograca31** [comment]
Extensions have access to the same window object as the tab? Right?
If so, you could just do `window.onbeforeunload = false;` before suspending that tab....

**Comment by AndrewJDR** [comment]
Ah, makes sense. I haven't done much extension development so I wasn't sure how it worked.

However, I just tried reproducing the issue and was not able to. So I may be barking up the wrong tree here. I'm now not sure what causes these pages to steal focus when they suspend. Hopefully I will find a reproduction case some day and fix it.

**Comment by AndrewJDR** [comment]
@rodrigograca31 I'm testing this change:
https://github.com/AndrewJDR/thegreatsuspender/commit/274ba3ccd4c0612c936ffa66ec102507b72ac685

Unfortunately, it doesn't seem to work for your example page:
https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_onbeforeunload
If I click run, then let the tab get suspended, it still pops up the dialog. Perhaps I've added my code in the wrong place. Feedback welcome.

**Comment by rodrigograca31** [comment]
This issue is from 2015... I don't remember the exact problem but:

If you go to: https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_onbeforeunload
And try to suspend it it shows an alert message....
(Chrome 58)

That probably also happens if the extension trys to suspend it after X minutes...
I think I've not experienced that in a while maybe it's already fixed but as I said this is from 2015.....

**Comment by deanoemcke** [comment]
I agree this is annoying. But I could not find a way to turn these popups off either.
Here is a working site that shows a popup on exit (assuming you type something in the text area first).
One hesitation I have about implementing such a feature (assuming I even could) would be that often these types of popups are there for a reason. Because you really dont want the user to lose a lot of typed information for example.

**Comment by AndrewJDR** [comment]
is it possible to detect whether a page has any onunload listeners registered?

If so, another idea would be to have an option to allow users to simply not suspend pages that have onunload/onbeforeunload listeners registered.
